lightning/src/blinded_path/payment.rs

```diff
@@ -0,0 +1,337 @@
+///
+/// [`BlindedHop`]: crate::blinded_path::BlindedHop
+pub struct PaymentRelay {
+	/// Number of blocks subtracted from an incoming HTLC's `cltv_expiry` for this [`BlindedHop`].
+	///
+	///[`BlindedHop`]: crate::blinded_path::BlindedHop
+	pub cltv_expiry_delta: u16,
+	/// Liquidity fee charged (in millionths of the amount transferred) for relaying a payment over
+	/// this [`BlindedHop`], (i.e., 10,000 is 1%).
+	///
+	///[`BlindedHop`]: crate::blinded_path::BlindedHop
+	pub fee_proportional_millionths: u32,
+	/// Base fee charged (in millisatoshi) for relaying a payment over this [`BlindedHop`].
+	///
+	///[`BlindedHop`]: crate::blinded_path::BlindedHop
+	pub fee_base_msat: u32,
+}
+
+/// Constraints for relaying over a given [`BlindedHop`].
+///
+/// [`BlindedHop`]: crate::blinded_path::BlindedHop
+pub struct PaymentConstraints {
+	/// The maximum total CLTV delta that is acceptable when relaying a payment over this
+	/// [`BlindedHop`].
+	///
+	///[`BlindedHop`]: crate::blinded_path::BlindedHop
+	pub max_cltv_expiry: u32,
+	/// The minimum value, in msat, that may be relayed over this [`BlindedHop`].
+	pub htlc_minimum_msat: u64,
+}
+
+impl Writeable for BlindedPaymentTlvs {
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
+		// TODO: write padding
+		match self {
+			Self::Forward { short_channel_id, payment_relay, payment_constraints, features } => {
+				encode_tlv_stream!(w, {
+					(2, short_channel_id, required),
+					(10, payment_relay, required),
+					(12, payment_constraints, required),
+					(14, features, required),
+				});
+			},
+			Self::Receive { payment_secret, payment_constraints, features } => {
+				encode_tlv_stream!(w, {
+					(6, payment_secret, required),
+					(12, payment_constraints, required),
+					(14, features, required),
+				});
+			}
+		}
+		Ok(())
+	}
+}
+
+impl Readable for BlindedPaymentTlvs {
+	fn read<R: io::Read>(r: &mut R) -> Result<Self, DecodeError> {
+		_init_and_decode_tlv_stream!(r, {
+			(1, _padding, option),
+			(2, scid, option),
+			(6, payment_secret, option),
+			(10, payment_relay, option),
+			(12, payment_constraints, required),
+			(14, features, required),
+		});
+		if let Some(short_channel_id) = scid {
+			if payment_secret.is_some() { return Err(DecodeError::InvalidValue) }
+			Ok(BlindedPaymentTlvs::Forward {
+				short_channel_id,
+				payment_relay: payment_relay.ok_or(DecodeError::InvalidValue)?,
+				payment_constraints: payment_constraints.0.unwrap(),
+				features: features.0.unwrap(),
+			})
+		} else {
+			if payment_relay.is_some() { return Err(DecodeError::InvalidValue) }
+			Ok(BlindedPaymentTlvs::Receive {
+				payment_secret: payment_secret.ok_or(DecodeError::InvalidValue)?,
+				payment_constraints: payment_constraints.0.unwrap(),
+				features: features.0.unwrap(),
+			})
+		}
+	}
+}
+
+/// Construct blinded payment hops for the given `unblinded_path`.
+pub(super) fn blinded_hops<T: secp256k1::Signing + secp256k1::Verification>(
```

Comment:
It'd be tricky for the onion message portion, because `ForwardTlvs` and `ReceiveTlvs` are different types so can't be part of the same `Vec`/`iter` unless we use `dyn`, which incurs a performance cost (here's my reference: https://stackoverflow.com/questions/27957103/how-do-i-create-a-heterogeneous-collection-of-objects).

Reply:
Good point, that simplified the OM-side code quite a bit! Thanks